keps/sig-apps/3939-allow-recreation-when-fully-terminated/README.md

```diff
@@ -0,0 +1,699 @@
+
+No
+
+#### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+For Job API, we are adding a enum field named `RecreatePodsWhen` which takes either a `TerminateOrFailed` or `Failed`
+
+- API type(s): enum
+- Estimated increase in size: 8B
+
+We are also added a status field for tracking terminating pods.
+
+- API type(s): int32
+- Estimated increase in size: 4B
+
+#### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+<!--
+Look at the [existing SLIs/SLOs].
+
+Think about adding additional work or introducing new steps in between
+(e.g. need to do X to start a container), etc. Please describe the details.
+
+[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
+-->
+
+#### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+N/A
+<!--
+Things to keep in mind include: additional in-memory state, additional
+non-trivial computations, excessive access to disks (including increased log
+volume), significant amount of data sent and/or received over network, etc.
+This through this both in small and large cases, again with respect to the
+[supported limits].
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+-->
+
+#### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?
+
+N/A
+<!--
+Focus not just on happy cases, but primarily on more pathological cases
+(e.g. probes taking a minute instead of milliseconds, failed pods consuming resources, etc.).
+If any of the resources can be exhausted, how this is mitigated with the existing limits
+(e.g. pods per node) or new limits added by this KEP?
+
+Are there any tests that were run/should be run to understand performance characteristics better
+and validate the declared limits?
+-->
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+#### How does this feature react if the API server and/or etcd is unavailable?
+
+No change from existing behavior of the Job controller.
+
+#### What are other known failure modes?
+
+#### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+- Initial KEP
+
+## Drawbacks
+
+Enabling this feature may have rollouts become slower.
+
+## Alternatives
+
+We discussed having this under the PodFailurePolicy but this is a more general idea than the PodFailurePolicy.
```

Comment:
Can you elaborate more, since I asked this above.

Reply:
I wrote a response in another place.  I'd like to hear api-review thoughts on this now or @mimowo?

TLDR was that we originally wanted to include it in retriable-kep but it was thought to be better brought up as a separate feature.  I thought that would mean we should decouple it from this API but I realize that I am not so sure anymore.

Reply:
I think the reason to split was to avoid the snowball effect when more loosely related goals are added to the pod failure policy KEP. In particular it introduces a new API field, so requires at least two releases (alpha and beta), which could prolong graduation of the pod failure policy unnecessarily.